flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/queryConfig.scala

```diff
@@ -19,9 +19,21 @@
 package org.apache.flink.table.api
 
 import _root_.java.io.Serializable
+
 import org.apache.flink.api.common.time.Time
 
-class QueryConfig private[table] extends Serializable {}
+class QueryConfig private[table] extends Serializable {
+}
+
+object QueryConfig {
+  def getQueryConfigFromTableEnv(tableEnv: TableEnvironment): QueryConfig = {
```

Comment:
The `queryConfig` method in `StreamTableEnvironment` does the same thing as the overridden one, just because the class return type, this makes the code duplicated.

There's another subclass `MockTableEnvironment`,  so the default method in `TableEnvironment` may be

```
private[flink] def queryConfig: QueryConfig = this match {
case _: BatchTableEnvironment => new BatchQueryConfig
case _: StreamTableEnvironment => new StreamQueryConfig
case _ => ???
}
```

If we decide to add an `queryConfig` method, I suggest make it abstract and  implements in every subclass.

Reply:
I agree with @lincoln-lil , we can define an abstract method in `TableEnvironment`  and implement it in `StreamTableEnvironment` which do not need to cast QueryConfig to StreamQueryConfig manually.

```
def queryConfig: StreamQueryConfig = new StreamQueryConfig
```

Reply:
The problem is that we should have the correct subtype of `QueryConfig` (not only the returned object but also the method signature) and the separation of the stream and batch APIs. The `StreamExecutionEnvironment.queryConfig` method must be typed to `StreamQueryConfig` to prevent casting. However, we don't want to have a method that is typed to `StreamQueryConfig` in the base class `TableEnvironment` because 1) this class should be independent from the streaming API and 2) we don't want a streaming method in the `BatchTableEnvironment`.

So, we can add a method to `TableEnvironment` that returns `QueryConfig` (depending on its own type a `StreamQueryConfig` or `BatchQueryConfig`) and override this method in `StreamTableEnvironment` such that it returns a `StreamQueryConfig` (and if later necessary in `BatchTableEnvironment` to return a `BatchQueryConfig`).

The duplicated code are < 5 lines and we have correct types everywhere.